pkg/ipam/datastore.go

```diff
@@ -46,6 +52,110 @@ func (c *CIDRResource) IsFull() bool {
 	return len(c.Used) == c.Size
 }
 
+func (c *CIDRResource) Contains(ip net.IP) bool {
+	if ip.To4() == nil {
+		return false
+	}
+
+	if c.CIDR != nil {
+		return c.CIDR.Contains(ip)
+	}
+
+	start := binary.BigEndian.Uint32(c.FirstIP.To4())
+	end := binary.BigEndian.Uint32(c.LastIP.To4())
+	value := binary.BigEndian.Uint32(ip.To4())
+	return start <= value && value <= end
+}
+
+func (c *CIDRResource) IntersectsWith(value *CIDRResource) bool {
```

Comment:
It looks like the implementation of MetalLB is more clear?
https://sourcegraph.com/github.com/metallb/metallb@b6678ebb6cb7b0ec591e88de9b34808d59212b1e/-/blob/internal/config/config.go#L455:6

Reply:
MetalLB uses the package `mikioh/ipaddr` to convert the IP range to multiple `net.IPNet`.  
And then compare these IPNets to the targets. I think this code is not very efficient.
For example, we set `192.168.0.1-192.168.0.10` to the address, and then the function `ipaddr.Summarize(start, end)` will retrun 5 IPNets 
```
192.168.0.1/32
192.168.0.2/31
192.168.0.4/30
192.168.0.8/31
192.168.0.10/32
```
In my way, I only compare 2 firstIP and 2 lastIP. 
But if we use MetalLB's implementation, the code may looks more simpler.  😎